service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateAzureNetworkStep.java

```diff
@@ -0,0 +1,141 @@
+              .withTag("workspaceId", resource.getWorkspaceId().toString())
+              .withTag("resourceId", resource.getResourceId().toString())
+              .defineRule("AllowHttpInComing")
+              .allowInbound()
+              .fromAddress("INTERNET")
+              .fromAnyPort()
+              .toAnyAddress()
+              .toPort(8080)
+              .withProtocol(SecurityRuleProtocol.TCP)
+              .attach()
+              .create(
+                  Defaults.buildContext(
+                      CreateNetworkSecurityGroupRequestData.builder()
+                          .setName(resource.getSubnetName())
+                          .setRegion(Region.fromName(resource.getRegion()))
+                          .setResourceGroupName(azureCloudContext.getAzureResourceGroupId())
+                          .setRules(Collections.emptyList())
+                          .build()));
+
+      computeManager
+          .networkManager()
+          .networks()
+          .define(resource.getNetworkName())
+          .withRegion(resource.getRegion())
+          .withExistingResourceGroup(azureCloudContext.getAzureResourceGroupId())
+          .withTag("workspaceId", resource.getWorkspaceId().toString())
+          .withTag("resourceId", resource.getResourceId().toString())
+          .withAddressSpace(resource.getAddressSpaceCidr())
+          .defineSubnet(resource.getSubnetName())
+          .withAddressPrefix(resource.getSubnetAddressCidr())
+          .withExistingNetworkSecurityGroup(subnetNsg)
+          .attach()
+          .create(
+              Defaults.buildContext(
+                  CreateNetworkRequestData.builder()
+                      .setName(resource.getNetworkName())
+                      .setResourceGroupName(azureCloudContext.getAzureResourceGroupId())
+                      .setRegion(Region.fromName(resource.getRegion()))
+                      .setSubnetName(resource.getSubnetName())
+                      .setNetworkSecurityGroup(subnetNsg)
+                      .setAddressPrefix(resource.getSubnetAddressCidr())
+                      .setAddressSpaceCidr(resource.getAddressSpaceCidr())
+                      .build()));
+    } catch (ManagementException e) {
+      // Stairway steps may run multiple times, so we may already have created this resource. In all
+      // other cases, surface the exception and attempt to retry.
+      // Azure error codes can be found here:
+      // https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/common-deployment-errors
+      if (StringUtils.equals(e.getValue().getCode(), "Conflict")) {
+        logger.info(
+            "Azure Network {} in managed resource group {} already exists",
+            resource.getNetworkName(),
+            azureCloudContext.getAzureResourceGroupId());
+        return StepResult.getStepResultSuccess();
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
```

Comment:
Are there classes of failures that should just cause an exit here?

Reply:
I think it would make sense to add `SubnetsNotInSameVnet` to the `CreateAzureNetworkStep`. A vm's network and subnet must be in the same virtual network based on [Azure docs](https://docs.microsoft.com/en-us/azure/azure-resource-manager/troubleshooting/common-deployment-errors) - I'll update this now. For `GetAzureNetworkStep` there's really only `ResourceNotFound` so I'll leave as is!